contracts/TokenIssuing.sol

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.6.0 <0.7.0;
+
+import "@openzeppelin/contracts/proxy/Initializable.sol";
+
+import "./common/Ownable.sol";
+import "./common/Pausable.sol";
+import "./interfaces/IRelay.sol";
+import "./interfaces/IERC20.sol";
+import "./interfaces/ISettingsRegistry.sol";
+import { ScaleStruct } from "./common/Scale.struct.sol";
+
+import "./common/Scale.sol";
+import "./common/SafeMath.sol";
+
+pragma experimental ABIEncoderV2;
+
+contract TokenIssuing is Ownable, Pausable, Initializable {
+
+    event MintRingEvent(address recipient, uint256 value, bytes32 accountId);
+    event MintKtonEvent(address recipient, uint256 value, bytes32 accountId);
+    event MintTokenEvent(address token, address recipient, uint256 value, bytes32 accountId);
+
+    bytes public storageKey;
+
+    ISettingsRegistry public registry;
+    IRelay public relay;
+
+    // Record the block height that has been verified
+    mapping(uint32 => bool) history;
+
+    function initialize(address _registry, address _relay, bytes memory _key) public initializer {
+        ownableConstructor();
+        pausableConstructor();
+
+        relay = IRelay(_relay);
+        registry = ISettingsRegistry(_registry);
+
+        storageKey = _key;
+    }
+
+    // The last step of the cross-chain of darwinia to ethereum, the user needs to collect some signatures and proofs after the darwinia network lock token.
+    // This call will append mmr root and verify mmr proot, events proof, and mint token by decoding events. If this mmr root is already included in the relay contract, the contract will skip verifying mmr root msg, saving gas.
+
+    // message - bytes4 prefix + uint32 mmr-index + bytes32 mmr-root
+    // signatures - the signatures for mmr-root msg
+    // root, MMRIndex - mmr root for the block
+    // blockNumber, blockHeader - The block where the lock token event occurred in darwinia network
+    // can be fetched by api.rpc.chain.getHeader('block hash') 
+    // peaks, siblings - mmr proof for the blockNumber
+    // eventsProofStr - mpt proof for events
+    // Vec<Vec<u8>> encoded by Scale codec
+    function appendRootAndVerifyProof(
+        bytes memory message,
+        bytes[] memory signatures,
+        bytes32 root,
+        uint32 MMRIndex,
+        uint32 blockNumber,
+        bytes memory blockHeader,
```

Comment:
bytes can use bytes32

Reply:
not a hash, the 'blockHeader' stores the entire block header content (Parent Hash + State Root + Extrinsics Root ...)